### PR TITLE
Fix collector config update

### LINKED ISSUE #460: Restart collector on config change
At the moment operator does not restart the collector when `spec.config` part of the CR changes (e.g. configure new processor). The operator could restart the collector if the config changes. 

The config update could be implemented on the collector itself, but that seems like a long way to go https://github.com/open-telemetry/opentelemetry-collector/issues/273.

### PR DESCRIPTION
**Pull request by mcariapas**
I added check sum to pod annotations. This will fix rollout when the collector config changes.
Yes, I know this is duplicated pull request with #506 
 but the other one has been forgotten for some time.
 
 Fixes #460

### CODE REVIEW

**Comment by mcariasap** [comment]
@pavolloffay Can you review it?

File: apis/v1alpha1/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -1,4 +1,3 @@
-//go:build !ignore_autogenerated
```
Comment (VineethReddy02) [inline apis/v1alpha1/zz_generated.deepcopy.go:1]: Why is this removed?

File: go.sum [hunk 2]
```diff
@@ -27,22 +27,30 @@ cloud.google.com/go/bigquery v1.3.0/go.mod h1:PjpwJnslEMmckchkHFfq+HTD2DmtT67aNF
 cloud.google.com/go/bigquery v1.4.0/go.mod h1:S8dzgnTigyfTmLBfrtrhyYhwRxG72rYxvftPBK2Dvzc=
 cloud.google.com/go/bigquery v1.5.0/go.mod h1:snEHRnqQbz117VIFhE8bmtwIDY80NLUZUMb4Nv6dBIg=
 cloud.google.com/go/bigquery v1.7.0/go.mod h1://okPTzCYNXSlb24MZs83e2Do+h+VXtc4gLoIoXIAPc=
+cloud.google.com/go/bigquery v1.8.0 h1:PQcPefKFdaIzjQFbiyOgAqyx8q5djaE7x9Sqe712DPA=
```
Comment (VineethReddy02) [inline go.sum]: I do not see any go import changes in this PR. Why do we need these changes?

File: apis/v1alpha1/zz_generated.deepcopy.go (hunk 1 above)
Comment (mcariapas) [reply apis/v1alpha1/zz_generated.deepcopy.go:1]: It was generated automatically....

File: go.sum (hunk 2 above)
Comment (mcariapas) [reply go.sum]: I don't know how go mode works in background, but when I tested it I downloaded all dependencies. I can remove these changes from the PR.

File: apis/v1alpha1/zz_generated.deepcopy.go (hunk 1 above)
Comment (mcariapas) [reply apis/v1alpha1/zz_generated.deepcopy.go:1]: I can remove it from the PR if it is necessary

File: pkg/collector/annotations.go [hunk 3]
```diff
@@ -37,10 +37,18 @@ func Annotations(instance v1alpha1.OpenTelemetryCollector) map[string]string {
...
+
+	if nil != instance.Spec.PodAnnotations {
+		for k, v := range instance.Spec.PodAnnotations {
+			podAnnotations[k] = v
+		}
+	}
+
 	// make sure sha256 for configMap is always calculated
 	annotations["opentelemetry-operator-config/sha256"] = getConfigMapSHA(instance.Spec.Config)
+	podAnnotations["opentelemetry-operator-config/sha256"] = getConfigMapSHA(instance.Spec.Config)
```
Comment (pavolloffay) [inline pkg/collector/annotations.go:49]: Get the hash only once and use it twice

File: pkg/collector/annotations.go [hunk 4]
```diff
@@ -22,9 +22,9 @@ import (
 )
 
 // Annotations return the annotations for OpenTelemetryCollector pod.
-func Annotations(instance v1alpha1.OpenTelemetryCollector) map[string]string {
+func Annotations(instance v1alpha1.OpenTelemetryCollector) (map[string]string, map[string]string) {
```
Comment (pavolloffay) [inline pkg/collector/annotations.go:25]: I liked more a separate function  `PodAnnotations` from the original PR. 

Please also update the godocs.

File: go.sum (hunk 2 above)
Comment (pavolloffay) [inline go.sum]: As @VineethReddy02 mentioned there should not be changes to this file as the PR does not change any dependencies.

**Review by pavolloffay — COMMENTED** [review]
Thanks for looking into this and it looks good to me ❤️ 

I would love to see e2e test for this change (something like this)
1. deploy collector as deployment
2. assert it is deployed
3. change config in CR - e.g. add new receiver
4. assert the collector pod restarted with the new changes - check if the new receiver port is exposed.

File: tests/e2e/smoke-restarting-deployment/00-errors.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: restarting-collector
+spec:
+  ports:
+  - targetPort: 4317
```
Comment (pavolloffay) [inline tests/e2e/smoke-restarting-deployment/00-errors.yaml:7]: Does this file define that the service should not expose 4317 port?

**Review by pavolloffay — APPROVED** [review]

File: tests/e2e/smoke-restarting-deployment/00-errors.yaml (hunk 5 above)
Comment (mcariapas) [reply tests/e2e/smoke-restarting-deployment/00-errors.yaml:7]: yes, because we use it when we change the configuration and we have to check if it is not already.

**Comment by mcariapas** [comment]
Can we merge it before the next release?

**Comment by pavolloffay** [comment]
yes, thanks for the PR :) 

cc) @jpkrohling @VineethReddy02 could you please merge?
